src/structures/CommandInteraction.js

```diff
@@ -0,0 +1,195 @@
+
+    /**
+     * Whether this interaction has already been replied to
+     * @type {boolean}
+     */
+    this.replied = false;
+
+    /**
+     * An associated webhook client, can be used to create deferred replies
+     * @type {WebhookClient}
+     */
+    this.webhook = new WebhookClient(this.applicationID, this.token, this.client.options);
+  }
+
+  /**
+   * The invoked application command, if it was fetched before
+   * @type {?ApplicationCommand}
+   */
+  get command() {
+    const id = this.commandID;
+    return this.guild?.commands.cache.get(id) ?? this.client.application.commands.cache.get(id) ?? null;
+  }
+
+  /**
+   * Defers the reply to this interaction.
+   * @param {boolean} [ephemeral] Whether the reply should be ephemeral
+   * @returns {Promise<void>}
+   */
+  async defer(ephemeral) {
+    if (this.deferred || this.replied) throw new Error('INTERACTION_ALREADY_REPLIED');
+    await this.client.api.interactions(this.id, this.token).callback.post({
+      data: {
+        type: InteractionResponseTypes.DEFERRED_CHANNEL_MESSAGE_WITH_SOURCE,
+        data: {
+          flags: ephemeral ? MessageFlags.FLAGS.EPHEMERAL : undefined,
+        },
+      },
+    });
+    this.deferred = true;
+  }
+
+  /**
+   * Options for a reply to an interaction.
+   * @property {boolean} [ephemeral] Whether the reply should be ephemeral
+   * @typedef {MessageOptions} InteractionReplyOptions
+   */
+
+  /**
+   * Creates a reply to this interaction.
+   * @param {string|APIMessage|MessageAdditions} content The content for the reply
+   * @param {InteractionReplyOptions} [options] Additional options for the reply
+   * @returns {Promise<void>}
+   */
+  async reply(content, options) {
+    if (this.deferred || this.replied) throw new Error('INTERACTION_ALREADY_REPLIED');
+    const apiMessage = content instanceof APIMessage ? content : APIMessage.create(this, content, options);
+    const { data, files } = await apiMessage.resolveData().resolveFiles();
+
+    await this.client.api.interactions(this.id, this.token).callback.post({
+      data: {
+        type: InteractionResponseTypes.CHANNEL_MESSAGE_WITH_SOURCE,
+        data,
+      },
+      files,
+    });
+    this.replied = true;
+  }
+
+  /**
+   * Fetches the initial reply to this interaction.
+   * @returns {Promise<Message|Object>}
+   */
+  async fetchReply() {
+    const raw = await this.client.api.webhooks(this.applicationID, this.token).messages('@original').get();
+    return this.channel?.messages.add(raw) ?? raw;
+  }
+
+  /**
+   * Edits the initial reply to this interaction.
+   * @param {string|APIMessage|MessageEmbed|MessageEmbed[]} content The new content for the message
+   * @param {WebhookEditMessageOptions} [options] The options to provide
+   * @returns {Promise<Message|Object>}
+   */
+  async editReply(content, options) {
+    const raw = await this.webhook.editMessage('@original', content, options);
+    return this.channel?.messages.add(raw) ?? raw;
```

Comment:
```suggestion
    return this.webhook.editMessage('@original', content, options);
```
`Webhook#editMessage()` already handles this, see [here](https://github.com/discordjs/discord.js/blob/master/src/structures/Webhook.js#L246-L254) for reference.

Reply:
Taken from the description:

> Returns the raw message data if the webhook was instantiated as a WebhookClient or if the channel is uncached, otherwise a Message will be returned

And since `interaction.webhook` is a WebhookClient, this should stay here

Reply:
Ah yeah sorry mb, I overlooked that. Nevertheless, this should still probably also play it safe and follow the same pattern by returning a patched clone to avoid race conditions with the gateway event (which could result in both params of `Client:messageUpdate` to be the updated message)

Reply:
on that note, why is it a WebhookClient? I do not see a benefit to having a separate client if it does not need, and there is a downside that is ratelimit handling being separated out (i.e. global ratelimits if hit are rediscovered on interactions).

And since we're on that note, the jsdoc references to WebhookClient#deleteMessage and WebhookClient#editMessage will not work, that class does not document the properties.

Reply:
Fixed the jsdoc references, the reason why there's a webhook client is so that the methods do not need to be re-implemented. While that'd be easily doable for `editReply()` and `deleteReply()`, the client can also be used for followup messages which would mean even more methods on the `CommandInteraction` class.